torchao/prototype/quantization/quantize_/workflows/float8/float8_semisparse_tensor.py

```diff
@@ -0,0 +1,232 @@
+    ):
+        super().__init__()
+        self.qdata = qdata  # dense fp8 for dequantization
+        self.qdata_compressed = qdata_compressed  # compressed for matmul
+        self.scale = scale
+        self.block_size = block_size
+        self.original_shape = original_shape
+        self.float8_dtype = float8_dtype
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}("
+            f"qdata_shape={self.qdata.shape}, {self.scale=}, "
+            f"{self.block_size=}, {self.shape=}, {self.device=}, {self.dtype=})"
+        )
+
+    @property
+    def qdata_fp8(self):
+        """For test compatibility"""
+        return self.qdata
+
+    @classmethod
+    def from_hp(
+        cls,
+        w: torch.Tensor,
+        block_size: list[int],
+        float8_dtype: torch.dtype = torch.float8_e4m3fn,
+    ):
+        if w.dim() != 2 or len(block_size) != 2:
+            raise ValueError("Expected 2D tensor and block_size length 2")
+
+        if not w.is_cuda:
+            raise ValueError("Semi-sparse layout requires CUDA tensors")
+
+        # Verify dimensions are compatible with 2:4 compression
+        rows, cols = w.shape
+        if rows % 32 != 0 or cols % 32 != 0:
+            raise ValueError(
+                "Tensor dimensions must be multiples of 32 for CUDA sparse compression"
+            )
+
+        # Validate block_size
+        if not all(bs > 0 for bs in block_size):
+            raise ValueError(f"block_size must be positive, got {block_size}")
+
+        if rows % block_size[0] != 0 or cols % block_size[1] != 0:
+            raise ValueError(
+                f"Dimensions {w.shape} must be divisible by block_size {block_size}"
+            )
+
+        # Apply 2:4 sparsity pruning
+        with torch.no_grad():
+            w_sparse = w.clone()
+
+        pruning_inds = w_sparse.abs().view(-1, 4).argsort(dim=1)[:, :2]
+        w_sparse.view(-1, 4).scatter_(1, pruning_inds, value=0)
+
+        # Check for all-zero (sparsity=1) tensor
+        if w_sparse.abs().max() == 0:
+            raise ValueError("Input tensor is all zeros after pruning")
+
+        scale = _choose_qparams_affine_floatx(w_sparse, ebits=4, mbits=3)
+
+        if not torch.isfinite(scale).all():
+            raise ValueError("Scale contains NaN or inf values")
+        if not (scale > 0).all():
+            raise ValueError(f"Scale contains non-positive values: min={scale.min()}")
+
+        scale_expanded = scale.unsqueeze(1)
+        scaled_data = w_sparse / scale_expanded
+        scaled_data = scaled_data.clamp(-448.0, 448.0)
+        fp8_data = scaled_data.to(float8_dtype).contiguous()
+
+        if not torch.isfinite(fp8_data.to(torch.float32)).all():
+            raise ValueError("fp8_data contains NaN/Inf after quantization")
+
+        # Store fp8 data in both dense and compressed formats
+        fp8_data_fp16 = fp8_data.to(torch.float16)
+
+        fp8_compressed = to_sparse_semi_structured(fp8_data_fp16)
```

Comment:
We should use the torchao cutlass packing kernels here, not the default torch ones: https://github.com/pytorch/ao/blob/315e9b4555f2245bb452b9e22f485419188ce8b4/torchao/dtypes/floatx/cutlass_semi_sparse_layout.py#L171